### PR TITLE
VACMS-3470: fix publish latest revision action

### PR DESCRIPTION
**Pull request by acrollet**
## Description

See #3470.

## Testing done

Manual testing, added unit test for new service.

## QA steps

As an admin user:
1. Open the Bulk Edit Content page (/admin/content/bulk?title=Psychology+Internships+and+Fellowships&type=health_care_region_detail_page&moderation_state=All&taxonomy_entity_index_tid_depth=All)
1. Open the "Psychology Internships and Fellowships" node in a new tab and save a draft with a new title
1. On the Bulk Edit Content page, select the "Psychology Internships and Fellowships" and apply the "Publish Latest Revision" action
1. Validate that:
- [x] The published revision has the new title

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [x] No announcement is needed for this code change. 
  - [x] Merge & carry on unburdened by announcements

### CODE REVIEW

**Comment by indytechcook** [comment]
It looks good code wise.  I nice refactor.  I'm waiting until test pass to do the review in devshop.

**Review by ndouglas — DISMISSED** [review]
I tested locally @indytechcook and it worked as expected.  Code LGTM too.

**Comment by acrollet** [comment]
@oksana-c pinging in case you'd like to validate this.

File: docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    if ($node instanceof RevisionLogInterface) {
+      $node->setRevisionCreationTime($this->datetimeTime->getRequestTime());
+      $node->setRevisionLogMessage('Bulk operation create archived revision');
+      $node->setRevisionUserId($this->currentUser->id());
+    }
+    $node->save();
+
+    $node = $this->entityTypeManager->getStorage('node')->load($node->id());
+    $node->set('moderation_state', 'draft');
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php:62]: is there a reason why we would want to create a `draft` after creating `archived` version?

File: docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+   *   The node.
+   */
+  public function unpublishCurrentRevision(NodeInterface $node) : NodeInterface {
+    $node->set('moderation_state', 'archived');
+
+    if ($node instanceof RevisionLogInterface) {
+      $node->setRevisionCreationTime($this->datetimeTime->getRequestTime());
+      $node->setRevisionLogMessage('Bulk operation create archived revision');
+      $node->setRevisionUserId($this->currentUser->id());
+    }
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php:58]: revision info is not applied on save. 

`$node instanceof RevisionLogInterface` - did you mean `NodeInterface`? do we need `if` statement here?

**Review by oksana-c — CHANGES_REQUESTED** [review]
attaching annotated screenshots. We may need a 16th min on this one today

### testing VBO on existing content


let's rename these two custom actions while we're at it

rename actions

----

### Observations while creating new content then running it thru Unpublish bulk action

Revisions in content region seems to align, although no revision messages are added and expect that author and timestamp are also taken from the last revision before the VBO was ran. 
In the right sidebar, the view shows an extra Draft revision, which is expected, because current logic in this PR saves Archived and then Draft revision. I can guess the rationale why Draft was added as a last revision, although I think we should not add Draft after Archived for Unpublish action, since users need to get accustomed to the concept of "Archived" and make a conscious decision on whether or not content gets resurrected as Draft after been Archived.
Screen Shot 2020-11-16 at 7 02 02 AM

strangely, Draft is displayed as last status in the bulk content view
Screen Shot 2020-11-16 at 7 02 42 AM

File: docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php (hunk 1 above)
Comment (acrollet) [reply docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php:62]: good question, I assumed that this was an original requirement.

File: docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php (hunk 2 above)
Comment (acrollet) [reply docroot/modules/custom/va_gov_bulk/src/Service/ModerationActions.php:58]: Good point, I think this action was originally meant to be available for all entities so I believe that was the reasoning behind `RevisionLogInterface`. However there were other dependencies on the object being a node, so I made the logic explicit and I think we can remove the `if` statement.

**Comment by acrollet** [comment]
@oksana-c let's definitely 16th minute this and get on the same page since we're talking about changing the spec. :-)

**Comment by acrollet** [comment]
I'm also seeing that the content creation time is shown in the 'Recent changes' view, which means that when a revision is published via bulk action, the date shown is when the draft was created rather than when it was published. Let's discuss whether to change that behavior as well.

**Comment by acrollet** [comment]
Capturing action items discussed in 16th minute:

- [ ] change "Unpublish current revision" to "Archive selected content (Unpublish) "
- [ ] Archive action should not also create a new draft revision
- [ ] change "Publish Latest Revision" to "Publish latest revision"
- [ ] update "Recent changes" view to display revision update timestamp instead of revision created timestamp

**Review by oksana-c — APPROVED** [review]
both actions work well. I'm opening a follow up issue to use Revision Created time instead of Changed for sorting the right sidebar "Recent changes" view, since it's definitely out of scope here.
✅ good to merge from me

**Comment by acrollet** [comment]
@indytechcook @ndouglas if I could get a re-pass on the CR from one of you that would be awesome, thanks :-)

**Comment by ndouglas** [comment]
@acrollet What's the timetable?  I should be able to get to this today if it's not _too_ rushed, but it might be a couple hours.

**Comment by acrollet** [comment]
I think we're past today's window to test and deploy, so any time today should be fine. cc @oksana-c

**Review by ndouglas — APPROVED** [review]
I confirmed that this worked as expected and reflected the update to the action name, etc.

**Comment by acrollet** [comment]
many thanks @ndouglas !

**Pull request merged**